test/puppetlabs/puppetdb/testutils/log.clj

```diff
@@ -0,0 +1,164 @@
+  "Runs body with the current appenders of the logger specified by
+  logger-id replaced by the specified appenders.  If logger-id is not
+  a class, it is converted via str."
+  [logger-id appenders & body]
+  `(#'puppetlabs.puppetdb.testutils.log/with-log-appenders-fn
+     ~logger-id ~appenders (fn [] ~@body)))
+
+(defmacro with-log-event-listener
+  "Calls (listen event) for each logger-id event produced during the
+  execution of the body.  If logger-id is not a class, it is converted
+  via str."
+  [logger-id listen & body]
+  ;; Specify the root logger via org.slf4j.Logger/ROOT_LOGGER_NAME.
+  `(with-additional-log-appenders ~logger-id
+     [(doto (#'puppetlabs.puppetdb.testutils.log/log-event-listener ~listen)
+        .start)]
+     (do ~@body)))
+
+(defmacro with-logging-to-atom
+  "Conjoins all logger-id events produced during the execution of the
+  body to the destination atom, which must contain a collection.  If
+  logger-id is not a class, it is converted via str."
+  [logger-id destination & body]
+  ;; Specify the root logger via org.slf4j.Logger/ROOT_LOGGER_NAME.
+  `(with-log-event-listener
+     ~logger-id
+     (fn [event#] (swap! ~destination conj event#))
+     ~@body))
+
+(defn- suppressing-appender
+  [log-path]
+  (let [pattern "%-4relative [%thread] %-5level %logger{35} - %msg%n"
+        context (LoggerFactory/getILoggerFactory)]
+    (doto (FileAppender.)
+      (.setFile log-path)
+      (.setAppend true)
+      (.setEncoder (doto (PatternLayoutEncoder.)
+                     (.setPattern pattern)
+                     (.setContext context)
+                     (.start)))
+      (.setContext context)
+      (.start))))
+
+(def ^:private annoying-peer-error-rx #"peer.*vm://localhost.*stopped")
+
+(defn notable-pdb-event? [event]
+  (and (.isGreaterOrEqual (.getLevel event) Level/ERROR)
+       (not (and (-> (.getFormattedMessage event)
+                     (.contains "queue connection error"))
+                 (when-let [cause (.getThrowableProxy event)]
+                   (re-find annoying-peer-error-rx (.getMessage cause)))))))
+
+(defn- with-log-suppressed-unless-notable-fn [notable-event? f]
+  (let [problem (atom false)
+        detector (doto (#'puppetlabs.puppetdb.testutils.log/log-event-listener
+                        (fn [event]
+                          (when (notable-event? event)
+                            (reset! problem true))))
+                   .start)
+        log-path (fs/absolute-path (temp-file "pdb-suppressed" ".log"))
+        appender (suppressing-appender log-path)]
+    (try
+      (with-log-appenders org.slf4j.Logger/ROOT_LOGGER_NAME
+        [appender detector]
+        (f))
+      (finally
+        (.stop appender)
+        (when @problem
+          (binding [*out* *err*]
+            (print (slurp log-path))
+            (println "From error log:" log-path)))
+        (when-not @problem (fs/delete log-path))))))
+
+(defmacro with-log-suppressed-unless-notable
```

Comment:
If I recall correctly, only if you establish the atom logger inside the body (since this macro uses with-log-appenders, which intentionally clobbers everything for the duration).